src/System.Buffers/src/System/Buffers/ArrayPoolBucket.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Provides a thread-safe bucket containing buffers that can be Rented and Returned as part 
+    /// of a buffer pool; it should not be used independent of the pool.
+    /// </summary>
+    internal sealed class ArrayPoolBucket<T>
+    {
+        private volatile int _index;
+        private T[][] _data;
+        private SpinLock _lock;
+
+        /// <summary>
+        /// Creates the pool with numberOfBuffers arrays where each buffer is of bufferLength length.
+        /// </summary>
+        internal ArrayPoolBucket(int bufferLength, int numberOfBuffers)
+        {
+            _index = 0;
+            _lock = new SpinLock();
+            _data = new T[numberOfBuffers][];
+            for (int i = 0; i < numberOfBuffers; i++)
+            {
+                _data[i] = new T[bufferLength];
+            }
```

Comment:
I had the following reasons behind this:

1) Since the bucket itself is lazy initialized, the only time the user will access the bucket will be when that data is actually needed. The configuration options will allow them to fine tune how many buffers they need, so having excess shouldn't be a big problem. In terms of the shared pool, having excess would be a good thing since each component will probably need a different number. 

2) Since the bucket itself is lazy initialized and, when it is initialized, creates all the buffers upfront, all calls to Rent (except when the bucket is empty) are constant time. I'd prefer to keep this functionality, if possible, since it provides a constant performance benefit; once you need a bucket, rents are guaranteed to be cheap as long as you stay within the bucket bounds.

3) This also keeps the bucket code simple :)

That being said, if you think that the memory pressure around these scenarios would be too much, I'm open to changing the functionality :)